src/Npgsql/Internal/NpgsqlReadBuffer.cs

```diff
@@ -209,56 +210,62 @@ internal Task Ensure(int count, bool async, bool readingNotifications)
+                    case IOException when (e.InnerException as SocketException)?.SocketErrorCode ==
                                             (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
                     {
                         Debug.Assert(e is OperationCanceledException ? async : !async);
 
-                        // When reading notifications (Wait), just throw TimeoutException or OperationCanceledException immediately.
-                        // Nothing to cancel, and no breaking of the connection.
-                        if (readingNotifications)
-                        {
-                            if (connector.UserCancellationRequested)
-                                throw;
-                            throw NpgsqlTimeoutException();
-                        }
-
-                        // If we should attempt PostgreSQL cancellation, do it the first time we get a timeout.
-                        // TODO: As an optimization, we can still attempt to send a cancellation request, but after that immediately break the connection
-                        if (connector.AttemptPostgresCancellation &&
-                            !connector.PostgresCancellationPerformed &&
-                            connector.PerformPostgresCancellation()
+                        var isStreamBroken = false;
 #if NETSTANDARD2_0
-                            // SslStream on .NET Framework treats any IOException (including timeouts) as fatal and may return garbage if reused.
-                            // To prevent this, we flow down and break the connection immediatelly.
-                            // See #4305.
-                            && !(connector.IsSecure && e is IOException)
+                        // SslStream on .NET Framework treats any IOException (including timeouts) as fatal and may
+                        // return garbage if reused. To prevent this, we flow down and break the connection immediately.
+                        // See #4305.
+                        isStreamBroken = connector.IsSecure && e is IOException;
 #endif
-                            )
+
+                        if (!isStreamBroken)
                         {
-                            // Note that if the cancellation timeout is negative, we flow down and break the connection immediately.
-                            var cancellationTimeout = connector.Settings.CancellationTimeout;
-                            if (cancellationTimeout >= 0)
+                            // When reading notifications (Wait), just throw TimeoutException or
+                            // OperationCanceledException immediately.
+                            // Nothing to cancel, and no breaking of the connection.
+                            if (readingNotifications)
+                                throw CreateException(connector);
```

Comment:
Note a small fix here: before, if a cancellation was requested, we just bubbled up the exception, unlike non-Wait where we would throw a new OperationCanceledException (presumably to get the right cancellation token). This PR aligns Wait with that.